src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,252 @@
+    active_trackers: AtomicUsize,
+    /// Index of the next free recorder.
+    next_idle: AtomicUsize,
+}
+
+impl insert::Lazy<Tracker> for InsertTracker {
+    fn value(&self) -> Tracker {
+        self.0.clone().tracker()
+    }
+}
+
+// ===== impl Scope =====
+
+impl Scope {
+    pub fn new() -> Self {
+        Scope(Arc::new(Shared::new()))
+    }
+
+    pub fn layer(&self) -> insert::Layer<InsertTracker, Tracker> {
+        insert::Layer::new(InsertTracker(self.0.clone()))
+    }
+}
+
+impl FmtMetric for Scope {
+    const KIND: &'static str = <Histogram<latency::Us> as FmtMetric>::KIND;
+
+    fn fmt_metric<N: fmt::Display>(&self, f: &mut fmt::Formatter<'_>, name: N) -> fmt::Result {
+        if let Ok(hist) = self.0.histogram.lock() {
+            hist.fmt_metric(f, name)?;
+        }
+        Ok(())
+    }
+
+    fn fmt_metric_labeled<N, L>(
+        &self,
+        f: &mut fmt::Formatter<'_>,
+        name: N,
+        labels: L,
+    ) -> fmt::Result
+    where
+        N: fmt::Display,
+        L: FmtLabels,
+    {
+        if let Ok(hist) = self.0.histogram.lock() {
+            hist.fmt_metric_labeled(f, name, labels)?;
+        }
+        Ok(())
+    }
+}
+
+// ===== impl InsertTracker =====
+
+impl Clone for Tracker {
+    fn clone(&self) -> Self {
+        self.shared.clone_tracker(self.idx);
+        Self {
+            shared: self.shared.clone(),
+            idx: self.idx,
+            t0: self.t0,
+        }
+    }
+}
+
+impl Drop for Tracker {
+    fn drop(&mut self) {
+        self.shared.drop_tracker(&*self);
+    }
+}
+
+impl Shared {
+    const INITIAL_RECORDERS: usize = 32;
+
+    fn new() -> Self {
+        let mut recorders = Vec::with_capacity(Self::INITIAL_RECORDERS);
+        Self::add_recorders(&mut recorders, Self::INITIAL_RECORDERS);
+        Self {
+            histogram: Mutex::new(Histogram::default()), // TODO(eliza): should we change the bounds here?
+            recorders: RwLock::new(recorders),
+            idle_head: AtomicUsize::new(0),
+        }
+    }
+
+    fn tracker(self: Arc<Self>) -> Tracker {
+        let t0 = Instant::now();
+        loop {
+            let idx = self.idle_head.load(Ordering::Relaxed);
+            // This is determined in a scope so that we can move `Self` into the
+            // new tracker without doing a second (unecessary) arc bump.
+            let acquired = {
+                // Do we have any free recorders remaining, or must we grow the
+                // slab?
+                let recorders = self
+                    .recorders
+                    .read()
+                    .ok()
+                    .filter(|recorders| idx < recorders.len())
+                    .unwrap_or_else(|| {
+                        // Slow path: if there are no free recorders in the
+                        // slab, extend it (acquiring a write lock temporarily).
+                        self.grow();
+                        self.recorders.read().unwrap()
```

Comment:
if we unwrap here, why not unwrap above (where you use `ok()`)?

Reply:
The `ok` above could also be an `unwrap`; the reason it's `ok()` is because we are optionally returning a recorder from the `filter`, and the `ok()` lets us nicely chain the option, without the `ok()` the code is a bit more complex. I can make this consistent if that's preferred.

Reply:
In general, I find the `filter` harder to read, and it's confusing because it at first seems like we're graceful to the lock being poisoned but, in fact, we just panic in the next access in that case.

```rust
let sz = self.recorders.read().unwrap().len();
if sz < idx {
  self.grow();
}
self.recorders.read().unwrap()
```

Is clearer about what it's doing, imo

Reply:
The intention was to avoid reacquiring the read lock in the fast path. The code you posted is definitely more readable but will always acquire the read lock twice, even when we don't need to acquire the write lock. It is entirely possible that the performance hit from reacquiring it is not significant enough to justify making the code less clear, though.